### PR TITLE
Add OpenTelemetry instrumentation

### PR DESCRIPTION
**Pull request by renchap**
This is based on the work from https://github.com/mastodon/mastodon/pull/20338 and https://github.com/mastodon/mastodon/pull/29016.

We want to deploy this to our staging environment to test how it behaves, and we can only build containers from PRs if the branch is part of the main repository, so I re-created the PR.

It mixes the good work from @robbkidd @julianocosta89

I went with `use_all` as advised by @robbkidd, if this load only the required gems then it should be equivalent, and is less verbose. I kept the documentation on how to disable a specific instrumentation.

The gems are now in a specific group, allowing people who do not want to install the OTEL gems to use `bundle --without opentelemetry`.

Once we have been able to test this, I will add back the collector config & build files from #29016

### CODE REVIEW

**Comment by renchap** [comment]
> * There's a typo in my email for the `Co-authored-by:`; it should end in `.org`.
>   * Or you can use my current work email `robbkidd@honeycomb.io` since I was working [the original contribution](https://github.com/mastodon/mastodon/pull/20338) officially on-the-clock.

Fixed, sorry. If needed, you can ping me on the Pollinators slack 

I just pushed a small change to add the current version in the `service.version` attribute, this should enable version-tracking features.

We also probably need to not send traces for the healthcheck endpoint.

We enabled sending OTEL traces to our Datadog account through the collector, and the span names are not very useful. Will try to see how we can improve those
image

File: Gemfile
Comment (mjankowski) [inline Gemfile:107]: Super-minor style thing here -- since every single gem in this group needs the option, we could put these lines into a `with_options require: false` block. Might reduce some noise and make every line marginally more readable with just the gem and version name.

Comment (renchap) [reply Gemfile:107]: I dont think you can use `with_options` in the Gemfile, because this is an Active Support method and it is not yet loaded

Comment (mjankowski) [reply Gemfile:107]: Oh right, duh. You are correct.

**Comment by renchap** [comment]
Pushed a new change which removes ActiveSupport, ActionPack and ActionView integrations, as well as the custom ActiveRecord one, because those are enabled by the Rails integration directly now.

**Comment by robbkidd** [comment]
> Pushed a new change which removes ActiveSupport, ActionPack and ActionView integrations, as well as the custom ActiveRecord one, because those are enabled by the Rails integration directly now.

The `opentelemetry-instrumentation-rails` metapackage will bring in and require `opentelemetry-instrumentation-active_record`.

⚠️ When I first attempted added OpenTelemetry, the direct ActiveRecord instrumentation (which uses prepending monkeypatches) [interacted badly](https://github.com/mastodon/mastodon/pull/20338#issuecomment-1311226122) with attr_encrypted (which at the time used alias_method monkeypatches). The custom ActiveRecord subscription was to avoid these colliding monkeys. 

I see attr_encrypted is still a dependency. It was [upgraded to 4.0.0](https://github.com/mastodon/mastodon/pull/24626) since I last tested things, but [4.0.0 still uses `alias_method`](https://github.com/attr-encrypted/attr_encrypted/blob/v4.0.0/lib/attr_encrypted/adapters/active_record.rb#L13).

I'll see if I see this bad interaction again locally!

File: config/initializers/opentelemetry.rb
Comment (robbkidd) [inline config/initializers/opentelemetry.rb:48]: I think we lost `c.use_all` somewhere along the way.

Here's a variant that configures the Rack instrumentation to not generate spans for the `/health` endpoint. It also needs to be set to not use Rack event subscriptions for instrumenting so it will use a Rack middleware instead. Acting as a wrapper, the middleware can better mark the whole trace context as ignorable. This is something we (the OTel instrumentation maintainers) should try to solve for with the Rack-event based instrumentation.

```suggestion

    c.use_all({
      'OpenTelemetry::Instrumentation::Rack' => {
        use_rack_events: false, # instead of events, use middleware; allows for untraced_endpoints to ignore child spans
        untraced_endpoints: ['/health'],
      },
    })
```

**Comment by mjankowski** [comment]
> I see attr_encrypted is still a dependency.

There is an approved but not yet merged PR which removes it - https://github.com/mastodon/mastodon/pull/28325 - part of larger migration to native AR encryption.

**Comment by robbkidd** [comment]
> ... part of larger migration to native AR encryption

❤️ This is a good and noble quest.

If the monkeypatches still conflict, the custom AS::Notification subscription can return as a temporary measure until native AR encryption lands.

**Comment by robbkidd** [comment]
📰 I'm running in docker compose locally with 41b1bff and I'm not seeing the infinite stack sadness of the colliding monkeypatches. So, that's a good sign.

File: config/initializers/opentelemetry.rb
Comment (renchap) [reply config/initializers/opentelemetry.rb:48]: Thanks, updated the PR with this!

**Review by julianocosta89 — APPROVED** [review]

**Comment by jgaskins** [comment]
Running this branch on [my instance](https://zomglol.wtf) (container built [here](https://hub.docker.com/r/jgaskins/mastodon/tags)) and seeing infinite recursion happening again. It's manifesting the same [as before](https://github.com/mastodon/mastodon/pull/20338#issuecomment-1312201367), in `AccountStat#reload`.

I haven't dug too deeply into it yet; I just got it built and deployed. Is it possible we still have conflicting monkeypatching strategies going on?

Screenshot of a trace on Honeycomb showing AccountStat reload calling itself recursively after an upsert

**Comment by mjankowski** [comment]
I haven't fully read the entire previous thing or dug deep into the arel/otel interaction here ... but FYI the PR which removes attr_encrypted has now been merged, but this branch is not yet rebased after that merge, so I'd expect whatever interactions existed there to still be present on this branch.

**Comment by renchap** [comment]
Rebased with the removal of `attr_encrypted`

**Comment by jgaskins** [comment]
> the PR which removes attr_encrypted has now been merged, but this branch is not yet rebased after that merge

Oooh, okay. I definitely misread the part up-thread.

I rebuilt the container with the rebase and deployed that and I haven't seen this issue since! Excellent work, folks!

**Review by ClearlyClaire — APPROVED** [review]
I'm somewhat uncomfortable with adding a large amount of dependencies, but this seems as self-contained as it gets, and the fact it is in an optional dependency group is appreciable.

**Comment by mjankowski** [comment]
I recall previous discussion about the OpenTelemetry integration replacing the nsa gem statsd support -- https://github.com/mastodon/mastodon/pull/29065#issuecomment-1929145755 -- on top of the initial work here, what things need to be done to enable that?

**Comment by renchap** [comment]
I would say that you can now get all the information from the NSA gem using OTEL (+ sidekiq-exporter maybe), + many more useful performance data, so I would be inclined to drop the NSA gem entirely?

**Comment by mjankowski** [comment]
Yeah ... I think that was the plan based on that original thread ... was just trying to confirm if we were there yet with just this PR, and it sounds like probably yes? I'll open PR to remove nsa and we can work through there if that's true.

**Comment by ClearlyClaire** [comment]
We already deprecated StatsD integration in 4.2.0, but we had no replacement available at the time. Furthermore, it was kinda buried in the [release notes](https://github.com/mastodon/mastodon/releases/v4.2.0)

**Comment by nomad-geek** [comment]
Submitted #30622 the other day which is somewhat related. If the STATSD_ADDR var is set and unavailable it will crash the containers.

**Pull request merged**
